runtime-modules/project-token/src/lib.rs

```diff
@@ -0,0 +1,611 @@
+
+        // ensure token validity
+        let _ = Self::ensure_token_exists(token_id)?;
+
+        // ensure src account id validity
+        let account_info = Self::ensure_account_data_exists(token_id, &who)?;
+
+        // ensure can freeze amount
+        ensure!(
+            account_info.free_balance >= amount,
+            Error::<T>::InsufficientFreeBalanceForReserving,
+        );
+
+        // == MUTATION SAFE ==
+
+        AccountInfoByTokenAndAccount::<T>::mutate(token_id, &who, |account_data| {
+            account_data.free_balance = account_data.free_balance.saturating_sub(amount);
+            account_data.reserved_balance = account_data.reserved_balance.saturating_add(amount);
+        });
+
+        Self::deposit_event(RawEvent::TokenAmountReservedFrom(token_id, who, amount));
+        Ok(())
+    }
+
+    /// Unreserve `amount` of token for `who`
+    /// Preconditions:
+    /// - `token_id` must id
+    /// - `who` must identify valid account for `token_id`
+    /// - `who` reserved balance must be greater than `amount`
+    ///
+    /// Postconditions:
+    /// - `who` free balance increased by `amount`
+    /// - `who` reserved balance decreased by `amount`
+    /// if `amount` is zero it is equivalent to a no-op
+    fn unreserve(token_id: T::TokenId, who: T::AccountId, amount: T::Balance) -> DispatchResult {
+        if amount.is_zero() {
+            return Ok(());
+        }
+
+        // ensure token validity
+        Self::ensure_token_exists(token_id).map(|_| ())?;
+
+        // ensure src account id validity
+        let account_info = Self::ensure_account_data_exists(token_id, &who)?;
+
+        // ensure can freeze amount
+        ensure!(
+            account_info.reserved_balance >= amount,
+            Error::<T>::InsufficientReservedBalance
+        );
+
+        // == MUTATION SAFE ==
+
+        AccountInfoByTokenAndAccount::<T>::mutate(token_id, &who, |account_data| {
+            account_data.free_balance = account_data.free_balance.saturating_add(amount);
+            account_data.reserved_balance = account_data.reserved_balance.saturating_sub(amount);
+        });
+
+        Self::deposit_event(RawEvent::TokenAmountUnreservedFrom(token_id, who, amount));
+
+        Ok(())
+    }
+
+    /// Retrieve reserved balance for token and account
+    /// Preconditions:
+    /// - `token_id` must be a valid token identifier
+    /// - `who` account id must exist for token
+    fn reserved_balance(
+        token_id: T::TokenId,
+        who: T::AccountId,
+    ) -> Result<T::Balance, DispatchError> {
+        let account_info = Self::ensure_account_data_exists(token_id, &who)?;
+
+        Ok(account_info.reserved_balance)
+    }
+
+    /// Retrieve total balance for token and account
+    /// Preconditions:
+    /// - `token_id` must be a valid token identifier
+    /// - `who` account id must exist for token
+    fn total_balance(token_id: T::TokenId, who: T::AccountId) -> Result<T::Balance, DispatchError> {
```

Comment:
What is this needed for?

Reply:
removed

Reply:
in b2726deded2bf71154e0e8e5e8ad01c4fbaaa257